# **Pandas - Python Library**

### **What is Pandas ?**

- **Pandas is a python library which helps in Data Analysis**
- **Used to analyze data**
- **Pandas has functions for analyzing, cleaning, exploring and manipulating data**

### **Why Use Pandas ?**

- **It helps to anlyze big data**
- **Makes conclusion on statistical theories**
- **Pandas can clean messy datasets**
- **Pandas methods can delete rows, clean data, rename columns, access specific data, filter data**

### **Installing Pandas**

## **- Pandas Series**

- **A pandas series is like column in a table**
- **It is one dimensional array holding data of any type**

## **- Pandas Dataframes**

- **Data sets in Pandas are usually multi-dimensional tables, called DataFrames.**
- **Series is like a column, a DataFrame is the whole table.**
- **Pandas dataframe is like 2 dimensional data structure, like two dimensional array, or table with columns and rows**
- **Dataframe is like table with columns and rows**

### **Import Pandas**

In [12]:
# importing pandas library so we can use its methods in our notebook
import pandas as pd     # here we create alias name for pandas library which is pd

### **Load Files into DataFrame**

#### **Pandas - Read CSV file**

In [15]:
# loading given dataset using read_csv() method of pandas 
# here df is a variable that stores data which we reading from this employee_csv dataset
df = pd.read_csv("C:\\Users\\HP\\Documents\\employee_dataset.csv")

## **Functions -**

### **1. shape** - Returns number of rows and columns

In [18]:
# how many rows and columns are there
df.shape

(100, 6)

- **In this dataset their are 100 rows and 6 columns**

### **2. head()** - Used to view rows

In [21]:
# display the first 3 rows
df.head(3)

EmployeeID   Name Department     Location  Age  Salary
0        1001    Ian  Marketing      Phoenix   60   42096
1        1002  Grace         HR  Los Angeles   45   68219
2        1003    Eva      Sales     New York   25   55736

In [22]:
# display first 5 rows
df.head()    # head() fuction display 5 rows in default 

EmployeeID     Name Department     Location  Age  Salary
0        1001      Ian  Marketing      Phoenix   60   42096
1        1002    Grace         HR  Los Angeles   45   68219
2        1003      Eva      Sales     New York   25   55736
3        1004  Charlie         HR      Chicago   44   57057
4        1005    Grace    Finance  Los Angeles   32   48955

### **3. columns** - Used to view column names

In [24]:
# display the column names
df.columns

Index(['EmployeeID', 'Name', 'Department', 'Location', 'Age', 'Salary'], dtype='object')

### **4. dtypes** - Used to view datatype of each column

In [26]:
# display datatypes of all columns
df.dtypes

EmployeeID     int64
Name          object
Department    object
Location      object
Age            int64
Salary         int64
dtype: object

1. **In this dataset EmployeeID, Age, Salary are of integer datatype**
2. **And Name, Department, Location are of object datatype**

### **5. isnull()** - For checking missing values

In [29]:
df.isnull()   # returns True where values are NAN and false otherwise and gives boolean output

EmployeeID   Name  Department  Location    Age  Salary
0        False  False       False     False  False   False
1        False  False       False     False  False   False
2        False  False       False     False  False   False
3        False  False       False     False  False   False
4        False  False       False     False  False   False
..         ...    ...         ...       ...    ...     ...
95       False  False       False     False  False   False
96       False  False       False     False  False   False
97       False  False       False     False  False   False
98       False  False       False     False  False   False
99       False  False       False     False  False   False

[100 rows x 6 columns]

In [30]:
# check for missing values 
df.isnull().sum()    # isnull() method checks if null and sum() method gives count of null values for each column

EmployeeID    0
Name          0
Department    0
Location      0
Age           0
Salary        0
dtype: int64

### **6. duplicated()** - Used to check duplicates

In [32]:
# check duplicate or not for each row
df.duplicated()    # check each row is duplicate or not and return boolean output

0     False
1     False
2     False
3     False
4     False
      ...  
95    False
96    False
97    False
98    False
99    False
Length: 100, dtype: bool

In [33]:
# get how many duplicate records in this dataset
df.duplicated().sum()   

0

In [34]:
df['Name'].duplicated().sum()     # duplicates in Name column

90

### **7. describe()** - Used to get statistical summary of dataframe

In [36]:
# get the basic statistical summary of numerical columns
df.describe()

EmployeeID         Age         Salary
count   100.000000  100.000000     100.000000
mean   1050.500000   39.660000   76764.470000
std      29.011492   11.701006   23632.927154
min    1001.000000   22.000000   40321.000000
25%    1025.750000   28.000000   56902.750000
50%    1050.500000   38.500000   71988.500000
75%    1075.250000   50.250000   96983.500000
max    1100.000000   60.000000  119368.000000

In [37]:
# to get statistical summary of numerical as well as categorical columns
df.describe(include = "all")

EmployeeID   Name Department Location         Age         Salary
count    100.000000    100        100      100  100.000000     100.000000
unique          NaN     10          5        5         NaN            NaN
top             NaN  Grace         IT  Chicago         NaN            NaN
freq            NaN     17         26       26         NaN            NaN
mean    1050.500000    NaN        NaN      NaN   39.660000   76764.470000
std       29.011492    NaN        NaN      NaN   11.701006   23632.927154
min     1001.000000    NaN        NaN      NaN   22.000000   40321.000000
25%     1025.750000    NaN        NaN      NaN   28.000000   56902.750000
50%     1050.500000    NaN        NaN      NaN   38.500000   71988.500000
75%     1075.250000    NaN        NaN      NaN   50.250000   96983.500000
max     1100.000000    NaN        NaN      NaN   60.000000  119368.000000

- **This dataset has 100 entries with no missing and duplicate values**
- **The average age of the employee is 39**
- **The minimum age of the employee is 22**
- **The maximum age of the employee is 60**
- **The average salary of the employee is 76764**
- **The minimum salary of the employee is 40321**
- **The maximum salary of the employee is 119368**

### **8. unique()** - Used to get unique values from a specific column 

In [40]:
# display unique values in a name column
df['Name'].unique()

array(['Ian', 'Grace', 'Eva', 'Charlie', 'Jane', 'Helen', 'Bob', 'David',
       'Alice', 'Frank'], dtype=object)

In [41]:
# display the unique values in department column
df['Department'].unique()

array(['Marketing', 'HR', 'Sales', 'Finance', 'IT'], dtype=object)

In [42]:
# display the unique values in Location column
df['Location'].unique()

array(['Phoenix', 'Los Angeles', 'New York', 'Chicago', 'Houston'],
      dtype=object)

### **9. value_counts()** - Count the number of occurences of each unique value in a column

In [44]:
# count the number of employees of location 
df['Location'].value_counts()

Location
Chicago        26
Houston        20
Phoenix        19
New York       18
Los Angeles    17
Name: count, dtype: int64

In [45]:
# count the number of employees of each department
df['Department'].value_counts()

Department
IT           26
Finance      24
HR           18
Marketing    17
Sales        15
Name: count, dtype: int64

In [46]:
# find no of employees from each name
df['Name'].value_counts()

Name
Grace      17
Ian        14
Helen      12
Jane       10
Bob        10
Eva         9
Charlie     9
David       8
Alice       7
Frank       4
Name: count, dtype: int64

### **10. nunique()** - Count no of unique values in a column

In [48]:
# find the number of unique locations
df['Location'].nunique()

5

In [49]:
# find the number of unique department
df['Department'].nunique()

5

In [50]:
# find the number of unique names
df['Name'].nunique()

10

### **11. info()** -  Used to get summary of dataframe

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   EmployeeID  100 non-null    int64 
 1   Name        100 non-null    object
 2   Department  100 non-null    object
 3   Location    100 non-null    object
 4   Age         100 non-null    int64 
 5   Salary      100 non-null    int64 
dtypes: int64(3), object(3)
memory usage: 4.8+ KB


### **12. max() / min()** - Returns maximum and minimum values

In [54]:
# what is maximum and minimum salary of employee
print("Max marks are :", df['Salary'].max())
print("Min marks are :", df['Salary'].min())

Max marks are : 119368
Min marks are : 40321


### **13. sum()** - Used to calculate sum of elements in a dataframe

In [56]:
df.head()

EmployeeID     Name Department     Location  Age  Salary
0        1001      Ian  Marketing      Phoenix   60   42096
1        1002    Grace         HR  Los Angeles   45   68219
2        1003      Eva      Sales     New York   25   55736
3        1004  Charlie         HR      Chicago   44   57057
4        1005    Grace    Finance  Los Angeles   32   48955

In [57]:
# how many employees are HR
(df['Department'] == "HR").sum()

18

In [58]:
# how many employee has location Phoenix
(df['Location'] == "Phoenix").sum()

19

In [59]:
# how many employees of age = 60
(df['Age'] == 60).sum()

3

### **14. len()** - length or entries in dataframe

In [61]:
len(df)

100

### **15. index()** - Returns the row index

In [63]:
df.index

RangeIndex(start=0, stop=100, step=1)

### **16. rename()** - rename the specified column or row

In [65]:
df.rename(columns = {"EmployeeID":"ID"}, inplace = True)

In [66]:
df.head()    # display first five rows

ID     Name Department     Location  Age  Salary
0  1001      Ian  Marketing      Phoenix   60   42096
1  1002    Grace         HR  Los Angeles   45   68219
2  1003      Eva      Sales     New York   25   55736
3  1004  Charlie         HR      Chicago   44   57057
4  1005    Grace    Finance  Los Angeles   32   48955

### **17. sort_values()** - Sort values

In [68]:
# sort the records by Salary in descending
df.sort_values(by = 'Salary', ascending = False).head()

ID   Name Department     Location  Age  Salary
54  1055  Helen         IT  Los Angeles   56  119368
24  1025   Jane         HR      Chicago   27  119311
51  1052    Ian         HR     New York   41  117742
37  1038    Eva         IT      Phoenix   22  117485
85  1086    Ian    Finance     New York   41  116158

In [69]:
# sort the records by age descending
df.sort_values(by = "Age", ascending = False).head()

ID   Name Department  Location  Age  Salary
0   1001    Ian  Marketing   Phoenix   60   42096
12  1013  Grace         HR   Houston   60   67425
71  1072   Jane      Sales  New York   60   61215
45  1046    Bob    Finance  New York   59   57004
18  1019  David    Finance   Chicago   59   59458

In [70]:
# sort the records by location
df.sort_values(by = "Location", ascending = True).head()

ID   Name Department Location  Age  Salary
41  1042  Grace         IT  Chicago   48   64992
67  1068    Eva         IT  Chicago   30  107477
66  1067   Jane      Sales  Chicago   28   99157
24  1025   Jane         HR  Chicago   27  119311
65  1066  Grace         IT  Chicago   33   66169

In [71]:
# axis = 0 for rows
# axis = 1 for columns

### **18. Filtering** - Selecting rows and columns that meets condition

In [73]:
# Show the details of employees from location Chicago
df[df['Location'] =='Chicago'].head()

ID     Name Department Location  Age  Salary
3   1004  Charlie         HR  Chicago   44   57057
6   1007    Helen    Finance  Chicago   51  110252
13  1014  Charlie    Finance  Chicago   33   72632
18  1019    David    Finance  Chicago   59   59458
24  1025     Jane         HR  Chicago   27  119311

In [74]:
# show the details of employees with age == 59
df[df['Age'] == 59]

ID   Name Department  Location  Age  Salary
18  1019  David    Finance   Chicago   59   59458
45  1046    Bob    Finance  New York   59   57004

In [75]:
# show the name of all employees having salary more than 80000
df[df['Salary'] > 80000].head()

ID     Name Department Location  Age  Salary
5   1006     Jane    Finance  Houston   36   96617
6   1007    Helen    Finance  Chicago   51  110252
8   1009  Charlie      Sales  Houston   26  110688
10  1011    Grace    Finance  Phoenix   24  110777
11  1012    Helen  Marketing  Houston   33   88183

In [76]:
# get records of age less than 40 
df[df['Age'] < 40].head()

ID     Name Department     Location  Age  Salary
2  1003      Eva      Sales     New York   25   55736
4  1005    Grace    Finance  Los Angeles   32   48955
5  1006     Jane    Finance      Houston   36   96617
8  1009  Charlie      Sales      Houston   26  110688
9  1010      Bob    Finance      Houston   25   59472

### **19. loc()** - Used to access rows and columns (Label - based - indexing)

In [78]:
# get name of the employees with name,location and age and return 5 records
df.loc[:4,['Name','Location','Age']]

Name     Location  Age
0      Ian      Phoenix   60
1    Grace  Los Angeles   45
2      Eva     New York   25
3  Charlie      Chicago   44
4    Grace  Los Angeles   32

In [79]:
# access all data
df.loc[:,:]

ID     Name Department     Location  Age  Salary
0   1001      Ian  Marketing      Phoenix   60   42096
1   1002    Grace         HR  Los Angeles   45   68219
2   1003      Eva      Sales     New York   25   55736
3   1004  Charlie         HR      Chicago   44   57057
4   1005    Grace    Finance  Los Angeles   32   48955
..   ...      ...        ...          ...  ...     ...
95  1096    Frank         IT      Chicago   51   73683
96  1097    Alice         IT      Phoenix   38   58905
97  1098      Ian  Marketing      Chicago   24   40321
98  1099    Alice         IT      Phoenix   44   40760
99  1100    David    Finance  Los Angeles   43  113139

[100 rows x 6 columns]

In [80]:
# get first 3 columns with all records
df.loc[:,['ID','Name','Department']].head()

ID     Name Department
0  1001      Ian  Marketing
1  1002    Grace         HR
2  1003      Eva      Sales
3  1004  Charlie         HR
4  1005    Grace    Finance

### **20. iloc()** - Used to access rows and columns (position - based - indexing)

In [82]:
# get info of 8 record 
df.iloc[8]   

ID               1009
Name          Charlie
Department      Sales
Location      Houston
Age                26
Salary         110688
Name: 8, dtype: object

In [83]:
# access records 6 with first five columns 
df.iloc[0:6, 0:5]

ID     Name Department     Location  Age
0  1001      Ian  Marketing      Phoenix   60
1  1002    Grace         HR  Los Angeles   45
2  1003      Eva      Sales     New York   25
3  1004  Charlie         HR      Chicago   44
4  1005    Grace    Finance  Los Angeles   32
5  1006     Jane    Finance      Houston   36

### **21. Add column**

In [85]:
df["Is_Senior"] = df["Age"] > 40    # add new column to dataset

In [86]:
df.head()

ID     Name Department     Location  Age  Salary  Is_Senior
0  1001      Ian  Marketing      Phoenix   60   42096       True
1  1002    Grace         HR  Los Angeles   45   68219       True
2  1003      Eva      Sales     New York   25   55736      False
3  1004  Charlie         HR      Chicago   44   57057       True
4  1005    Grace    Finance  Los Angeles   32   48955      False

### **22. Drop Column** - Used to drop the column

In [88]:
df.drop("Is_Senior", axis = 1, inplace = True)

In [89]:
df.head()

ID     Name Department     Location  Age  Salary
0  1001      Ian  Marketing      Phoenix   60   42096
1  1002    Grace         HR  Los Angeles   45   68219
2  1003      Eva      Sales     New York   25   55736
3  1004  Charlie         HR      Chicago   44   57057
4  1005    Grace    Finance  Los Angeles   32   48955

### **23. Rank()** - Rank values in a column. Assigning numerical position based on ranks

In [91]:
df["Salary"].rank(ascending = False).head()   # index is not changed

0    96.0
1    53.0
2    80.0
3    74.0
4    90.0
Name: Salary, dtype: float64

### **24. astype()** - Converts the datatype of column

In [93]:
df["Salary"].astype(float).head()

0    42096.0
1    68219.0
2    55736.0
3    57057.0
4    48955.0
Name: Salary, dtype: float64

### **25. apply()** - Used to apply function to each row or column 

In [95]:
df["Is_Senior"] = df["Age"].apply(lambda x : "Senior" if x > 40 else "Junior")

In [96]:
df.head()

ID     Name Department     Location  Age  Salary Is_Senior
0  1001      Ian  Marketing      Phoenix   60   42096    Senior
1  1002    Grace         HR  Los Angeles   45   68219    Senior
2  1003      Eva      Sales     New York   25   55736    Junior
3  1004  Charlie         HR      Chicago   44   57057    Senior
4  1005    Grace    Finance  Los Angeles   32   48955    Junior

### **26. query()** - Uses sql like queries to filter data 

In [98]:
df.query("Department == 'Sales' and Salary > 80000 and Is_Senior == 'Junior'").head()

ID     Name Department  Location  Age  Salary Is_Senior
8   1009  Charlie      Sales   Houston   26  110688    Junior
17  1018      Bob      Sales  New York   35   98083    Junior
19  1020    Grace      Sales  New York   34   99134    Junior
44  1045    Helen      Sales   Chicago   25  104759    Junior
47  1048      Ian      Sales   Phoenix   24  112395    Junior

### **27. pivot_table()** - Summarizes data by using aggregating functions

In [100]:
df.pivot_table(values = "Salary", index = "Location", columns = "Department", aggfunc = "mean")    

Department        Finance            HR            IT  Marketing     Sales
Location                                                                  
Chicago      66507.166667  80530.000000  69681.777778    40321.0   75299.0
Houston      81342.833333  62348.333333  77889.200000    73169.2  110688.0
Los Angeles  93108.250000  83061.333333  74021.666667    89086.4   75851.0
New York     86439.500000  89389.500000  84160.750000    67760.5   78542.0
Phoenix      79288.000000  69842.600000  81355.800000    52552.5   74755.0

### **28. tail()** - Returns last five rows

In [102]:
df.tail()

ID   Name Department     Location  Age  Salary Is_Senior
95  1096  Frank         IT      Chicago   51   73683    Senior
96  1097  Alice         IT      Phoenix   38   58905    Junior
97  1098    Ian  Marketing      Chicago   24   40321    Junior
98  1099  Alice         IT      Phoenix   44   40760    Senior
99  1100  David    Finance  Los Angeles   43  113139    Senior

### **29. group_by()** - Used to group data based on one or more columns and then apply a aggregate function 

In [104]:
# what is the average salary for each department
df.groupby('Department')['Salary'].max()

Department
Finance      116158
HR           119311
IT           119368
Marketing    114295
Sales        112395
Name: Salary, dtype: int64

In [105]:
df.groupby('Location')['Salary'].min()

Location
Chicago        40321
Houston        49841
Los Angeles    48955
New York       53636
Phoenix        40374
Name: Salary, dtype: int64

### **30. nlargest** - Top n rows with highest values

In [107]:
df.nlargest(5,"Salary")  

ID   Name Department     Location  Age  Salary Is_Senior
54  1055  Helen         IT  Los Angeles   56  119368    Senior
24  1025   Jane         HR      Chicago   27  119311    Junior
51  1052    Ian         HR     New York   41  117742    Senior
37  1038    Eva         IT      Phoenix   22  117485    Junior
85  1086    Ian    Finance     New York   41  116158    Senior

### **31. nsmallest** - Top n rows with smallest values

In [173]:
df.nsmallest(5,"Salary")

ID   Name Department Location  Age  Salary Is_Senior
97  1098    Ian  Marketing  Chicago   24   40321    Junior
49  1050   Jane         HR  Phoenix   40   40374    Junior
98  1099  Alice         IT  Phoenix   44   40760    Senior
70  1071    Bob         IT  Chicago   57   41974    Senior
0   1001    Ian  Marketing  Phoenix   60   42096    Senior